src/fsharp/LowerStateMachines.fs

```diff
@@ -0,0 +1,895 @@
+            if args.Length < macroParams.Length then 
+                //warning(Error(FSComp.SR.stateMachineMacroUnderapplied(), m))
+                None
+            else
+                let nowArgs, laterArgs = List.splitAt macroParams.Length args
+                let expandedExpr = MakeApplicationAndBetaReduce g (macroVal2, (tyOfExpr g macroVal2), [], nowArgs, m)
+                if sm_verbose then printfn "reduced application f = %A nowArgs= %A --> %A" macroVal2 nowArgs expandedExpr
+                if isNil laterArgs then 
+                    Some expandedExpr 
+                else
+                    if sm_verbose then printfn "application was partial, reducing further args %A" laterArgs
+                    TryReduceApp env expandedExpr laterArgs
+
+        | Expr.Let (bind, bodyExpr, m, _) -> 
+            match TryReduceApp env bodyExpr args with 
+            | Some bodyExpr2 -> Some (mkLetBind m bind bodyExpr2)
+            | None -> None
+
+        | Expr.LetRec (binds, bodyExpr, m, _) ->
+            match TryReduceApp env bodyExpr args with 
+            | Some bodyExpr2 -> Some (mkLetRecBinds m binds bodyExpr2)
+            | None -> None
+
+        | Expr.Sequential (x1, bodyExpr, sp, ty, m) -> 
+            match TryReduceApp env bodyExpr args with 
+            | Some bodyExpr2 -> Some (Expr.Sequential (x1, bodyExpr2, sp, ty, m))
+            | None -> None
+
+        // This construct arises from the 'mkDefault' in the 'Throw' case of an incomplete pattern match
+        | Expr.Const (Const.Zero, m, ty) -> 
+            Some (Expr.Const (Const.Zero, m, ty))
+
+        | Expr.Match (spBind, exprm, dtree, targets, m, ty) ->
+            let mutable newTyOpt = None
+            let targets2 = 
+                targets |> Array.choose (fun (TTarget(vs, targetExpr, spTarget, flags)) -> 
+                    // Incomplete exception matching expressions give rise to targets with I_throw. 
+                    // and System.Runtime.ExceptionServices.ExceptionDispatchInfo::Throw(...)
+                    // 
+                    // Keep these in the residue.
+                    //
+                    // In theory the type of the expression should be adjusted but code generation doesn't record the 
+                    // type in the IL
+                    let targetExpr2Opt = 
+                        match targetExpr, newTyOpt with 
+                        | Expr.Op (TOp.ILAsm ([ AbstractIL.IL.I_throw ], [_oldTy]), a, b, c), Some newTy -> 
+                            let targetExpr2 = Expr.Op (TOp.ILAsm ([ AbstractIL.IL.I_throw ], [newTy]), a, b, c) 
+                            Some targetExpr2
+                        | Expr.Sequential ((Expr.Op (TOp.ILCall ( _, _, _, _, _, _, _, ilMethodRef, _, _, _), _, _, _) as e1), Expr.Const (Const.Zero, m, _oldTy), a, b, c), Some newTy  when ilMethodRef.Name = "Throw" -> 
+                            let targetExpr2 = Expr.Sequential (e1, Expr.Const (Const.Zero, m, newTy), a, b, c)
```

Comment:
Hmmmm...  In theory yes.  Would it make any actual difference to call stacks etc? Or just smaller IL?

Reply:
To native stacks it would, printed traces would be fine due to the special casing of edi boundaries in IAsyncStateMachine. That change didn't get ported back to netfx though, if we still care at all about it.

Reply:
Hmmmm I see.  I will leave this for now, it's a possible later improvement.